### PR TITLE
Migrate project directory structure for codeworld-server

### PR DESCRIPTION
**Pull request by nixorn**
Foundation for  #827

### CODE REVIEW

**Comment by nixorn** [comment]
It is possible to get copy of production `data\haskell\projects` or `data\codeworld\projects` dir? It would be good to test migration locally.

**Comment by nixorn** [comment]
Locally tested moving/creating/deleting of directories and projects, as well as sharing directories.

File: codeworld-server/src/Migration.hs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+main :: IO ()
+main = do
+    print "Running migration of codeworld-server project structure."
+    alreadyDone <- isProjectsStructureCorrect
+    if alreadyDone then print "Migration already done."
+        else do
+            migrateMode (BuildMode "haskell")
+            migrateMode (BuildMode "codeworld")
+            print "Succsessfully migrated."
```
Comment (cdsmith) [inline codeworld-server/src/Migration.hs]: [nit] spelling error

File: codeworld-server/src/Migration.hs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    userDirs <- getAllUserDirs bm
+    mapM_ migrateProjectStructure userDirs
+
+main :: IO ()
+main = do
+    print "Running migration of codeworld-server project structure."
+    alreadyDone <- isProjectsStructureCorrect
+    if alreadyDone then print "Migration already done."
+        else do
+            migrateMode (BuildMode "haskell")
```
Comment (cdsmith) [inline codeworld-server/src/Migration.hs]: Please handle `BuildMode "blocklyXML"` as well

File: codeworld-server/src/Migration.hs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+isProjectsStructureCorrect :: IO Bool
+isProjectsStructureCorrect = do
+    let cwDir = projectRootDir (BuildMode "codeworld")
+        haskellDir = projectRootDir (BuildMode "haskell")
+    cwTree <- getDirectory cwDir
+    haskellTree <- getDirectory haskellDir
+    return $ checkForPrefDirs haskellTree && checkForPrefDirs cwTree
+    where
+        checkForPrefDirs :: FSTree -> Bool
+        checkForPrefDirs tr = null $ filter ((\x -> 3 == length x)) $ concatMap splitPath $ flatten tr
```
Comment (cdsmith) [inline codeworld-server/src/Migration.hs]: [nit] unnecessary second set of parens

File: codeworld-server/src/Main.hs [hunk 4]
```diff
@@ -345,39 +324,13 @@ moveProjectHandler = private $ \userId ctx -> do
...
-                    (getDirectoryContents
-                         (dropFileName $ moveFromDir </> projectFile projectId))
-            liftIO $
-                if empty
-                    then removeDirectoryIfExists
-                             (dropFileName $
-                              moveFromDir </> projectFile projectId)
-                    else removeFileIfExists $
-                         moveFromDir </> projectFile projectId
+            liftIO $ removeFileIfExists $ moveFromDir </> projectFile projectId
```
Comment (cdsmith) [inline codeworld-server/src/Main.hs]: Isn't this just `liftIO $ removeFileIfExists file` ?

File: codeworld-server/src/Migration.hs (hunk 1 above)
Comment (nixorn) [reply codeworld-server/src/Migration.hs]: done

File: codeworld-server/src/Migration.hs (hunk 2 above)
Comment (nixorn) [reply codeworld-server/src/Migration.hs]: done

File: codeworld-server/src/Migration.hs (hunk 3 above)
Comment (nixorn) [reply codeworld-server/src/Migration.hs]: done

File: codeworld-server/src/Main.hs (hunk 4 above)
Comment (nixorn) [reply codeworld-server/src/Main.hs]: yes, fixed.

File: codeworld-server/src/Migration.hs (hunk 2 above)
Comment (nixorn) [reply codeworld-server/src/Migration.hs]: ...But not checked how it works.

**Comment by nixorn** [comment]
Anyway before deploying this need to have backup of all production content and be ready to revert / rollback files.

File: codeworld-server/src/Migration.hs (hunk 2 above)
Comment (nixorn) [reply codeworld-server/src/Migration.hs]: Looks working.

**Comment by cdsmith** [comment]
Okay, I can snapshot the production data for testing.  (No, I can't send it to you, since it contains user data.)  I'll do that, and merge if it looks good.

**Comment by nixorn** [comment]
Also, I forgot to test "Sharing a folder, and accepting the shared folder."  yesterday. And looks I can not test it with local auth, because local auth disables on page reload, so I can not follow link being authorized.

**Comment by cdsmith** [comment]
Sorry for the delay; I've been overloaded with things to do. I plan to push this to a snapshot of the prod server today and do a bunch of manual testing. If that goes well, I'll merge tonight.

**Comment by nixorn** [comment]
No worries, I working on #827 on another branch.

**Pull request merged**